# Lesson 2: Intermediate Python

[View this lesson on DataCamp](https://learn.datacamp.com/courses/intermediate-python-for-data-science)

## Chapter 1: Matplotlib
This chapter introduces the basics of Matplotlib - how to make, configure, and show plots
### Lineplots

| |  |
|------------------ |----------------|
|`plt.plot(x,y)`    |create lineplot |
|`plt.show()`       |show plot       |
|`plt.title()`      |set plot title  |
|`plt.xlabel()`     |set x-axis label|
|`plt.ylabel()`     |set y-axis label|
|`plt.xscale('log')`|set x-axis to log scale|
|`plt.xticks()`     |define x-axis increments|
|`plt.yticks()`     |define y-axis increments|

### Scatterplots
Same as above, but with the scatter method
`plt.scatter(x,y)`

|Argument        |Function           |
|----------------|-------------------|
|`s=`            |size; sets dot size|
|`c=`            |colour, sets dot color|
|`alpha=`         |set data point transparency|

### Histograms 

|||
|--------------|----------------|
|`plt.hist()`  |create histogram, the `bins=` argument allows you to specify the number of bins|
|`plt.clf()`   |clean up plot a plot that's already been created so you can start fresh|

## Chapter 2: Dictionaries & pandas

### Dictionaries
Dictionaries are immutable, and composed of {'key':value} pairs

In [2]:
maritimes = {'NS':'Halifax', 'NB':'Fredricton', 'PEI':'Charlottetown'}

Calling `.keys()` prints all keys from a specified dictionary

In [3]:
maritimes.keys()

dict_keys(['NS', 'NB', 'PEI'])

Calling dict['key'] will return the value that corresponds with that key

In [4]:
maritimes['NS']

'Halifax'

Values can be added to dictionaries

In [5]:
maritimes['NFLD'] = 'St. Johns'
print(maritimes)

{'NS': 'Halifax', 'NB': 'Fredricton', 'PEI': 'Charlottetown', 'NFLD': 'St. Johns'}


...and deleted using `del()`

In [6]:
del(maritimes['NFLD'])
print(maritimes)

{'NS': 'Halifax', 'NB': 'Fredricton', 'PEI': 'Charlottetown'}


### pandas DataFrames

Pandas dataframes are a way of storing tabular data. While we could store a data table in a 2D NumPy array (as we previously learned), a pandas dataframe has a number of advantages. It's more like an Excel spreadsheet, in that it can have column names (labels) and row indexes (the equivalent of column names, but for rows). As well, pandas offers a variety of methods that make it easier and more intuitive to work with data than a NumPy array.

To import the pandas package we use the convention:

In [7]:
import pandas as pd

To convert an object, such as a NumPy array, to a pandas DataFrame, we use pd.DataFrame():

In [8]:
# Create np array
import numpy as np

x = np.array(([1,2,3,4],[5,6,7,8]))
print(x)

[[1 2 3 4]
 [5 6 7 8]]


In [9]:
# convert to DataFrame
df = pd.DataFrame(x)
df

0  1  2  3
0  1  2  3  4
1  5  6  7  8

Note above the pretty formatting that Jupyter applies when viewing a DataFrame. That alone makes it easier to work with than a NumPy array!

Note above that the column labels and indices are by default just sequential numbers, starting from zero as is the norm in Python.

Often we'll be reading data from a file rather than creating DataFrames from scratch. To load the contents of a CSV file into a DataFrame, use `pd.to_csv()`. Here we'll load a small CSV file containing 10 reaction times (RTs):

In [11]:
df = pd.read_csv('./3/rt_data.csv')
df

Trial_Num        RT
0          1  0.611653
1          2  0.007066
2          3  0.023062
3          4  0.524775
4          5  0.399861
5          6  0.046666
6          7  0.973756
7          8  0.232771
8          9  0.090606
9         10  0.618386

```
df = pd.read_csv('./rt_data.csv')
df
```

| |Trial_Num | RT | 
| --- | --- | --- |
|**0** | 1 | 0.611653|
|**1** | 2 | 0.007066|
|**2** | 3 | 0.023062|
|**3** | 4 | 0.524775|
|**4** | 5 | 0.399861|
|**5** | 6 | 0.046666|
|**6** | 7 | 0.973756|
|**7** | 8 | 0.232771|
|**8** | 9 | 0.090606|
|**9** | 10 | 0.618386|



Note above that the first imported column is `Trial_Num`, which starts at 1, but the index, by default, starts from zero. We can tell pandas that we want the trial number to act as the index, which could make our life easier later on when working with this data:

```
df = pd.read_csv('./rt_data.csv', index_col='Trial_Num')
df
```

| Trial_Num |  RT | 
| --- | --- |
|**1** | 0.611653|
|**2** | 0.007066|
|**3** | 0.023062|
|**4** | 0.524775|
|**5** | 0.399861|
|**6** | 0.046666|
|**7** | 0.973756|
|**8** | 0.232771|
|**9** | 0.090606|
|**10** | 0.618386|

### Slicing pandas dataframes
dataframe[#:#] allows you to slice specified rows based on integer index

|Slicing with `.loc` |                   |
|--------------------|-------------------|
|`.loc['']`         |slice specified row based on string index --> pandas series|
|`.loc[['']]`       |slice specified row based on string index --> pandas dataframe|
|`.loc[[''], ['']]` |slice specified rows and columns --> pandas sub-dataframe|
|`.loc[:, ' ']`     |slice all rows and specified columns --> pandas series|
|`.loc[:, [' ']]`   |slice all rows and specified columns --> pandas sub-dataframe|


|Slicing with `.iloc` |                   |
|--------------------|-------------------|
|`.iloc[#]`          |slice specified row based on integer index|
|`.iloc[[#]]`        |slice specified row based on unterger index --> pandas dataframe|
|`.iloc[[#], [#]]`   |slice specified rows and columns --> pandas sub-dataframe|
|`.iloc[:, #]`       |slice all rows and specified columns --> pandas series|
|`.iloc[:, [#]]`     |slice all rows and specified columns --> pandas sub-dataframe|

## Chapter 3: Logic, Control Flow, & Filtering

### Comparison Operators 
|               |               |
|---------------|---------------|
|`==`           |"are equal"    |
|`<` and `>`    |"is less than" and "is greater than" |
|`<=` and `>=`  |"less than or equal to" and "greater than or equal to"|
|`!=`           |"is not equal to"|

### Boolean Operators
|Operator          |Function       |
|------------------|---------------|
|and               |require 2 or more specified objects|
|or                |require at least one specified object|
|not               |anything but specified object|
|`np.logical_and()`|"and" statement, can be passed multiple comparison criteria|
|`np.logical_or()` |"or" statement, can be passed multiple comparison criteria|
|`np.logical_not()`|"not" statement, can be passed multiple comparison criteria|

### Conditional Statements

```
if condition :
     expression
elif condition : 
     expression
else :
     expression
```

`if` - check if initial specified condition holds True. If so, perform specified expression(s).

`elif` - (else if) if the `if` condition is False, check whether this secondary condition holds True. If so, perform secondary specified expression(s)

`else` - if all `if` or `elif` conditions are False, perform specified expression. 

Note that `elif` and `else` are *not required* in `if` statements. When you write an `if` statement with no `elif` oe `else` options, and the `if` statement is not True, then Python simply moves on as if nothing ever happened.

<div class="alert alert-block alert-info">
Always be sure your `if`, `elif`, and `else` lines end with colons!
</div>

In [9]:
x = 11

if x < 10: 
    print("Smaller than 10")
else:
    print("Larger than 10")

Larger than 10


A most complex example, to tell the season based on the month:

In [10]:
month = 'January'

if month == 'June' or month == 'July' or month == 'August':
    print('Summer!')
elif month == 'September' or month == 'October' or month == 'November':
    print("Fall")
elif month == 'December' or month == 'January' or month == 'February':
    print("Winter")
else:
    print("Spring")


Winter


## Chapter 4: Loops

### `while` Loops 
While loops will excecute the code many times, as long as the coniditon is true ("as long as x is true, do y")

Format:

```
while condition:
    expression
```



Count to 10, by twos:

In [11]:
x = 0

while x <= 10:
    print(x)
    x = x + 2

0
2
4
6
8
10


The code below stops once `x` becomes less than 6

In [12]:
x = 20

while x > 6:
    x = x/2
    print(x)

10.0
5.0


### `for` Loops
For loops are used to iterate over data types, and perform functions on each element

General Format:

```
list = [x, y, z]
for item in list :
      expression
```
      
See the example using the list 'odd' below

In [13]:
odd = [1, 3, 5, 7, 9]
for value in odd:
    print(value)

1
3
5
7
9


Use `enumerate()` to access index information (number of the current item in the sequence) during iteration:

In [14]:
names = ['Xhelal', 'Erika', 'Hezekiah', 'Antonia', 'Yaser']

# Just print the names
for x in names:
    print(x)

Xhelal
Erika
Hezekiah
Antonia
Yaser


In [15]:
# Enumerate and show the names
for value in enumerate(names):
    print(value)

(0, 'Xhelal')
(1, 'Erika')
(2, 'Hezekiah')
(3, 'Antonia')
(4, 'Yaser')


#### To loop over a list of lists:

In [16]:
lol = [['something', 4.2], ['something_else', 5.4]]

# Each list in our list has two entries; 
# if we just specify one variable to iterate over, we get each whole sub-list:
for this in lol :
      print(this)

['something', 4.2]
['something_else', 5.4]


<div class="alert alert-block alert-info">
Note above that when we iterate through a list (or dictionary, or array, as you'll see below), we make a new "dummy" variable name (and *iterator* to use inside the list for each entry from the list. In the example above we called the iterator `this`.
</div>

In [17]:
# But if we specify two variables, it assigns one to each entry within the sub-lists:
for this, that in lol:
    print(this)

something
something_else


In [18]:
for this, that in lol:
    print(that)

4.2
5.4


#### To loop over a dictionary

Remember that dicts are key:value pairs. Python iterates over the *keys* of dicts:

In [19]:
names_ages = {'Xhelal':22, 'Erika':25}

for name in names_ages :
    print(name)
#     print(names_ages[name])

Xhelal
Erika


If you want the *values* instead, you need to use indexing, where you use the iterator as the index to the original dict name:

In [20]:
for name in names_ages:
    print(names_ages[name])

22
25


Print key then value:

In [21]:
for name in names_ages:
    print(name)
    print(names_ages[name])

Xhelal
22
Erika
25


#### To loop over a numpy array:

1D arrays:

In [22]:
import numpy as np

odds = np.array([1, 3, 5, 7, 9])

for x in odds : 
    print(x + 100)

101
103
105
107
109


2D arrays:

In [23]:
odds_evens = np.array([[1,2], [3,4], [5,6], [7,8], [9,10]])
print(odds_evens)

[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]]


In [24]:
for x in np.nditer(odds_evens):
    print(x * 100)

100
200
300
400
500
600
700
800
900
1000


#### To loop over a dataframe

The beauty of pandas DataFrames is that they provide methods that automatically iterate over rows or columns, and they do so in ways that are much more efficient (and elegant code) than useing a `for` loop. So in most cases you should first consider how do do what you want without a `for` loop. But, if you really want to, here is how. 

Recall the DataFrame we created above, called `df`:


In [25]:
df

0  1  2  3
0  1  2  3  4
1  5  6  7  8

Let's multiply each value in each row by 1000:

In [26]:
# Original data is in seconds; print out RTs in milliseconds instead:
for index, row in df.iterrows() :
    print(row * 1000)

0    1000
1    2000
2    3000
3    4000
Name: 0, dtype: int64
0    5000
1    6000
2    7000
3    8000
Name: 1, dtype: int64


## Chapter 5: Hacker Statistics

### Random number generators

Generate a random number:

In [27]:
import numpy as np

np.random.rand()

0.9308370014914281

Run it again, and we'll get a differnt number (which is the point of 'random', right?)

In [28]:
np.random.rand()

0.8993407189765018

By default, the output of `np.random.rand()` is a random number between 0 and 1, drawn from a **uniform distribution** — in other words, any value is as likely as any other value in the range of 0–1. (One could also specify a different distribution, like a normal distribution with a mean of 0.5, where the value was most likely to be close to the centre of the distribution, i.e., close to 0.5).

If you want a random number in a different range, such as 0-100, you can simply multiply the result:

In [29]:
np.random.rand() * 100

8.751149805970782

However, if you want to specify a less rounded range, like, say, 78–94, and don't want to figure out the multiplier to do that, there is a "convenience function" that will allow you to specify the range. This is, arguably, more pythonic becuase it is more explicit/obvious to someone else reading your code:

In [30]:
np.random.randint(78, 94)

79

If you want more than one random number, you can generate a list (1d NumPy array) by giving the command a single numeric argument:

In [31]:
np.random.rand(5) 

array([0.4555505 , 0.37411796, 0.60489071, 0.81488882, 0.80393836])

You can generate a 2d NumPy array by giving it two arguments in the familiar (row, column) format:

In [32]:
np.random.rand(5, 2) 

array([[0.51237886, 0.97476443],
       [0.9821499 , 0.60054192],
       [0.71667158, 0.85983853],
       [0.45392903, 0.01051222],
       [0.73135922, 0.60583858]])

You can't chain the `rand()` and `randint()` commands, but the `randint()` takes an additional `size=` argument that allows you to specify the number/shape of the outputs:

In [33]:
np.random.randint(10,100, size=10)

array([80, 67, 34, 59, 59, 36, 24, 79, 27, 53])

In [34]:
np.random.randint(10,100, size=(5, 2)) # note size is a tuple

array([[27, 28],
       [68, 76],
       [44, 96],
       [88, 48],
       [37, 27]])

### Seeds

Sometimes, however, we want to generate the same random number every time. This may seem counter-intuitive, but it may be useful, for example if we want to write reproducible code that generates the same output every time it's run. One scientific application of this is simulations: sometimes, scientists want to create simulated data and test their algortihms on that, rather than on real data. They might want to do this becuase simulated data has known properties (e.g., you know the "truth"), whereas in real data, the point of analysis is often to identify certain properties that are not known. So if you are trying a new analysis method, and you don't know what the "true" result is because you're working with real data, it can be hard to validate the algorithm. If you use simulated data, you can get around that. 

As a simple example, we might want to simulate some human reaction time (RT) data. We might want to specify that it has a particular mean, but the simulated data would be a set of randomly-generated numbers, tha, when averaged, produced that mean. If we wanted to write some Python code in a Jupyter notebook to run this simulation, we might want it to generate the same simulated data each time — especially if this was example code that we were sharing with other scientists who would want to reproduce our results. In this case, we would want to specify the random **seed** — the number that "seeds" the random number generator to produce numebr or set of numbers.

Below we specify a seed of 42 ([the answer to the Ultimate Question of Life, the Universe, and Everything](https://en.wikipedia.org/wiki/42_(number)#The_Hitchhiker's_Guide_to_the_Galaxy)), and show that each time we do so, then call `np.random.rand()`, we get the same random number:

In [35]:
np.random.seed(42)
np.random.rand()

0.3745401188473625

In [36]:
np.random.seed(42)
np.random.rand()

0.3745401188473625

In [37]:
np.random.seed(42)
np.random.rand()

0.3745401188473625

Importantly though, we do need to explicitly call `np.random.seed(42)` each time we want to generate that same random number, otherwise the seed will reset the next time we run the command:

In [38]:
np.random.seed(42)
print(np.random.rand())
# Remember we need to "print" outputs if we want more than just the last command's
# output in a Jupyter notebook)

0.3745401188473625


In [39]:
print(np.random.rand())

0.9507143064099162


In [40]:
print(np.random.rand())

0.7319939418114051


In [41]:
np.random.seed(42)
print(np.random.rand())

0.3745401188473625


In [42]:
np.random.seed(42)
print(np.random.rand())

0.3745401188473625
